### ISSUE TITLE
Current workflow for releasing new banners

### ISSUE THREAD

**Issue by rowanmanning** [Financial-Times/n-messaging-client#213]
Currently we're rolling out major versions of all Origami components, progress with this work can be found in [this Google Sheet](https://docs.google.com/spreadsheets/d/1cmxGQp5ZgtAwSkS8Rww_4vk8xvN-GB9iAhNQf9iwcYs/edit#gid=1320807547). This project _has_ been updated and is now on version 4.x, however the applications which use it haven't been released with the updated version yet. These applications are:

- [ ] [next-article-page](https://github.com/Financial-Times/next-article-page)
- [ ] [next-front-page](https://github.com/Financial-Times/next-front-page)
- [ ] [next-stream-page](https://github.com/Financial-Times/next-stream-page)

What this means is that new banners have to be created on both the `master` branch (which is version 4.x), and the `v3.x` branch. They will also need to be tested on each of these branches. I'm going to outline the process for doing this below:

## Setup

I recommend re-cloning this repository so that you have two local copies. This will allow you to run both the 4.x and 3.x demos. Assuming you already have a local copy set up on the `master` branch, you can set up a 3.x branch like this:

```bash
git clone https://github.com/Financial-Times/n-messaging-client.git n-messaging-client-3x && cd n-messaging-client-3x && git checkout v3.x
```

Then you can run `make install` etc. If you want to run the demos for 4.x and 3.x in parallel, you will need to change the port for one of them. You can do this by prepending the demo command with a `PORT` environment variable:

```bash
PORT=1234 make demo
```

## Creating a banner

To create a banner for now, we're going to need two separate PRs which are checked out from the relevant base branch (`master` for v4.x, and `v3.x` for 3.x). I recommend working from `master` first, then porting changes to the older branch.

### Creating your 4.x banner

1. Checkout `master` and pull changes so you're up-to-date
2. Copy existing banner files to get you started, and do your work
3. Commit changes and open a new Pull Request against the `master` branch
4. Once the PR is merged, you can create a new release on GitHub. This will need to be a `v4.x.x` release and based on the `master` branch (the default):<br/><img width="466" alt="Screenshot 2019-12-04 at 09 01 03" src="https://user-images.githubusercontent.com/138944/70128442-b6604000-1674-11ea-86c0-ded931f77573.png">
5. Now for any of the next applications above which are checked (and _only_ those that are checked), you can update their `n-messaging-client` dependency to the new release you just created.

### Creating your 3.x banner

1. Checkout `v3.x` and pull changes so you're up-to-date
2. Copy your work from the `master` branch as a starting point
3. Make changes from the major Origami migrations, or the banner will not work in v3.x. These changes should be made in order, and applied only to the files that you're adding as part of the new banner. These will nearly always be `.scss`, `.html`, or `.js` files.
	1. Replace any reference to `n-messaging-client-messaging-banner` with `n-messaging-banner`
	2. Replace any reference to `o-banner` with `n-messaging-banner`
	3. Replace any reference to `n-messaging-client-alert-banner` with `n-alert-banner`
	4. Replace any reference to `o-message` with `n-alert-banner`
	5. Replace any reference to `oColorsByName` with `oColorsGetPaletteColor`
	6. Replace any reference to `oSpacingByName`, this will need some more work translating spacing names to old o-typography scales:
		- `oSpacingByName('s1')` becomes `oTypographySpacingSize($units: 1)`
		- `oSpacingByName('s2')` becomes `oTypographySpacingSize($units: 2)`
		- `oSpacingByName('s4')` becomes `oTypographySpacingSize($units: 4)`
		- `oSpacingByName('s6')` becomes `oTypographySpacingSize($units: 6)`
		- `oSpacingByName('m12')` becomes `oTypographySpacingSize($units: 10)`
	7. If there other Origami mixins used, you may have issues. At this point test that the demo builds, and if it doesn't then speak to a member of the Origami team referencing this document. We'll be happy to help
4. Commit changes and open a new Pull Request against the `v3.x` branch
5. Once the PR is merged, you can create a new release on GitHub. This will need to be a `v3.x.x` release and based on the `v3.x` branch. Use the dropdown to select this branch:<br/><img width="462" alt="Screenshot 2019-12-04 at 09 01 24" src="https://user-images.githubusercontent.com/138944/70128752-574efb00-1675-11ea-9274-095f0267a095.png">
6. Now for any of the next applications above which are **not** checked (and _only_ those that are not checked), you can update their `n-messaging-client` dependency to the new release you just created.
